src/ray/object_manager/pull_manager.cc

```diff
@@ -122,57 +156,87 @@ void PullManager::UpdatePullsBasedOnAvailableMemory(size_t num_bytes_available)
-    if (next_request_it == pull_request_bundles_.end()) {
-      // No requests are active. Get the first request in the queue.
-      next_request_it = pull_request_bundles_.begin();
-    } else {
-      next_request_it++;
+  std::unordered_set<ObjectID> object_ids_to_cancel;
+  // If there are any worker requests queued, try to activate them. Since we
+  // prioritize worker requests over task requests, task requests will be
+  // canceled as necessary to make space. We may exit this block over capacity
+  // if we run out of task requests to cancel, but this will be remedied later
+  // by canceling worker requests.
+  bool worker_requests_remaining = !worker_request_bundles_.empty();
+  while (worker_requests_remaining) {
+    worker_requests_remaining = ActivateNextPullBundleRequest(
+        worker_request_bundles_, &highest_worker_req_id_being_pulled_, &objects_to_pull);
+    while (num_bytes_being_pulled_ > num_bytes_available_) {
+      // Activating the last worker request put us over the available memory.
+      // Try to cancel task requests to make space.
+      if (highest_task_req_id_being_pulled_ == 0) {
+        // There are no task requests to cancel. Stop activating worker
+        // requests because we are already over capacity.
+        worker_requests_remaining = false;
+        break;
+      }
+
+      // Cancel a task request to make space.
+      RAY_LOG(DEBUG) << "Deactivating task args request "
+                     << highest_task_req_id_being_pulled_
+                     << " num bytes being pulled: " << num_bytes_being_pulled_
+                     << " num bytes available: " << num_bytes_available_;
+      const auto last_request_it =
+          task_argument_bundles_.find(highest_task_req_id_being_pulled_);
+      RAY_CHECK(last_request_it != task_argument_bundles_.end());
+      DeactivatePullBundleRequest(task_argument_bundles_, last_request_it,
+                                  &highest_task_req_id_being_pulled_,
+                                  &object_ids_to_cancel);
     }
+  }
 
-    if (next_request_it == pull_request_bundles_.end()) {
-      // No requests in the queue.
+  // If we still have capacity after processing worker requests, try to
+  // activate task requests.
+  while (num_bytes_being_pulled_ < num_bytes_available_) {
+    if (!ActivateNextPullBundleRequest(task_argument_bundles_,
+                                       &highest_task_req_id_being_pulled_,
+                                       &objects_to_pull)) {
       break;
     }
+  }
 
-    if (next_request_it->second.num_object_sizes_missing > 0) {
-      // There is at least one object size missing. We should not activate the
-      // bundle, since it may put us over the available capacity.
+  // While we are over capacity, deactivate requests starting from the back of
+  // the task request queue.
+  while (num_bytes_being_pulled_ > num_bytes_available_) {
+    if (highest_task_req_id_being_pulled_ == 0) {
```

Comment:
```suggestion
    if (highest_task_req_id_being_pulled_ <= 0) {
```
Consider this defensive change